procedures: Add docs about Open VSX on-premises (#2947)
codenvy/che-docs: modules/administration-guide/examples/snip_che-deploy-open-vsx-using-oc.adoc

* Installed `oc` and git tools.

> * Installed Podman.

Reviewer: do we need podman for downstream as well?

Author: No, we don’t need it. The image will be built and published on registry.redhat.io

* Log in to the OpenShift cluster where a {prod-short} is deployed as an administrator.
+
[TIP]
====

`$ oc login pass:c,a,q[{prod-url}] --username=__<my_user>__`

====

.Procedure

. Create a new OpenShift project for Open VSX.
+
[subs="+attributes,+quotes"]
----
oc new-project openvsx
----
. Clone Open VSX sources.
+
Get Open VSX sources and navigate to the OpenShift deploy directory:
+
[bash,subs="verbatim",options="nowrap"]
----
git clone https://github.com/svor/openvsx.git &&
cd openvsx &&
git checkout sv-update-os-deployment &&
cd deploy/openshift
----

. Prepare Open VSX server image.
+
The next step is to build and publish the Open VSX server image. To do this, run the following commands:
+
[bash,subs="verbatim",options="nowrap"]
----
export REGISTRY=quay.io &&
export NAMESPACE=myuser &&
export OPENVSX_SERVER_IMAGE_NAME=openvsx-server &&
export OPENVSX_VERSION=v0.27.0 &&
export OPENVSX_SERVER_IMAGE=${REGISTRY}/${NAMESPACE}/${OPENVSX_SERVER_IMAGE_NAME}:${OPENVSX_VERSION} &&
podman build -t "${OPENVSX_SERVER_IMAGE}" --build-arg "OPENVSX_VERSION=${OPENVSX_VERSION}" -f openvsx.Dockerfile . &&
podman login "${REGISTRY}" &&

> podman push "${OPENVSX_SERVER_IMAGE}"

Reviewer: those intrustions will be upstream-only right?

Author: Right. For downstream, there’s no need to build an image - the one built by Konflux will be used. Here’s the appropriate documentation for downstream: https://gitlab.cee.redhat.com/red-hat-developers-documentation/devspaces-documentation/-/merge_requests/131/diffs#ac357933a4d4693b96c1ee65e6657831686a2e76_0_1

Reviewer: verified on Developer Sandbox / workspaces.openshift.com

<img width="700" height="397" alt="Screenshot 2025-08-18 at 12 50 32" src="https://github.com/user-attachments/assets/ef8d8de7-e971-4854-871f-42c8d9e594f3" />

Reviewer: @svor that's a separate discussion, but I do believe in the blog post we should showcase the workspaces.openshift.com for build / deploy / update flow

----
+
[IMPORTANT]
====

> Make sure to configure the correct Docker registry and namespace for your image. Additionally, ensure that the image is publicly accessible so that your OpenShift cluster can pull it without requiring authentication.

Reviewer: ```suggestion
Make sure to configure the correct container registry and namespace for your image. Additionally, ensure that the image is publicly accessible and can be pulled without authentication.
```

====

. Prepare Open VSX CLI image.
+
Build and publish the Open VSX CLI image. To do this, run the following commands:
+
[bash,subs="verbatim",options="nowrap"]
----
export OPENVSX_CLI_IMAGE_NAME=openvsx-cli
export OPENVSX_CLI_IMAGE=${REGISTRY}/${NAMESPACE}/${OPENVSX_CLI_IMAGE_NAME} &&
podman build -t "${OPENVSX_CLI_IMAGE}" -f cli.Dockerfile . &&
podman push "${OPENVSX_CLI_IMAGE}"
----
+
[IMPORTANT]

> ====

Reviewer: wondering if we should publish an image on https://quay.io/ as well?

Author: Do you mean whether we need to publish the cli.Dockerfile, or are you asking if we should have prebuilt images for the OpenVSX CLI and OpenVSX available in the registry to avoid manual build/publish steps?

- Yes we need to publish images OpenVSX CLI and OpenVSX on the registry to reuse them when deploying OpneVSX
- For now, I wouldn’t publish these images ourselves (maybe need another issue for that, but for now I think we should be good with the current instructions)

Reviewer: @svor is it expected that cli would be tagged as latest unlike server?

```
openshift (sv-update-os-deployment) $ podman images
REPOSITORY                                         TAG              IMAGE ID      CREATED         SIZE
quay.io/ibuziuk/openvsx-cli                        latest           773693613f90  42 seconds ago  493 MB
quay.io/ibuziuk/openvsx-server                     v0.27.0          aad2c97f3413  3 minutes ago   815 MB
<none>                                             <none>           27da0b2d0568  4 minutes ago   1.05 GB
ghcr.io/eclipse/openvsx-server                     v0.27.0          d1c3bf2d33e7  5 weeks ago     809 MB
registry.access.redhat.com/ubi9                    9.5-1730489303   a60dfdf85e5e  9 months ago    235 MB
registry.access.redhat.com/ubi9/nodejs-20-minimal  1-63.1725851021  4c40309014ae  11 months ago   210 MB
```

Author: It should not affect anything, but it's better to have a tag. So, I've added tag for cli image as well

Ensure that the image is publicly accessible and can be pulled without authentication.

====

. Deploy Open VSX.
+
[bash,subs="verbatim",options="nowrap"]
----
oc process -f openvsx-deployment.yml \
-p OPENVSX_SERVER_IMAGE="${OPENVSX_SERVER_IMAGE}" \
-p OPENVSX_CLI_IMAGE="${OPENVSX_CLI_IMAGE}" \
| oc apply -f -

> ----

Reviewer: verified:

```
namespace/openvsx configured
persistentvolumeclaim/postgres-pvc created
deployment.apps/postgresql created
service/postgresql created
deployment.apps/elasticsearch created
service/elasticsearch created
persistentvolumeclaim/extensions-pvc created
deployment.apps/openvsx-server created
service/openvsx-server created
route.route.openshift.io/internal created
secret/github-oauth created
secret/ovsx-pat created
deployment.apps/ovsx-cli created
```

. Verify that all pods in the `openvsx` namespace are running and ready. It might take a few minutes for all pods to become ready. Run the following command:
+
[bash,subs="verbatim",options="nowrap"]
----
oc get pods -n openvsx \
  -o jsonpath='{range .items[*]}{@.metadata.name}{"\t"}{@.status.phase}{"\t"}{.status.containerStatuses[*].ready}{"\n"}{end}'

> ----

Reviewer: nice, verified via UI as well:

<img width="1620" height="580" alt="Screenshot 2025-08-18 at 13 00 06" src="https://github.com/user-attachments/assets/570bdd03-5105-46c6-ba48-a17ad046282b" />

. Add Open VSX user with PAT to the database.
+
Find PostgreSQL pod:
+
[bash,subs="verbatim",options="nowrap"]
----
export POSTGRESQL_POD_NAME=$(oc get pods -n openvsx \
   -o jsonpath="{.items[*].metadata.name}" | tr ' ' '\n' | grep '^postgresql' | head -n 1)
----
+
Insert username into OpenVSX database:
+
[bash,subs="verbatim",options="nowrap"]
----
oc exec -n openvsx "$POSTGRESQL_POD_NAME" -- bash -c \
   "psql -d openvsx -c \"INSERT INTO user_data (id, login_name, role) VALUES (1001, 'eclipse-che', 'privileged');\""
----
+
Insert user PAT into OpenVSX database:
+
[bash,subs="verbatim",options="nowrap"]
----
oc exec -n openvsx "$POSTGRESQL_POD_NAME" -- bash -c \
   "psql -d openvsx -c \"INSERT INTO personal_access_token (id, user_data, value, active, created_timestamp, accessed_timestamp, description) VALUES (1001, 1001, 'eclipse_che_token', true, current_timestamp, current_timestamp, 'extensions publisher');\""
----

> . Configure {prod-short} to use the internal Open VSX.

Reviewer: maybe we should write a bit about `openVSXURL` property, or at least point to / cross-link https://eclipse.dev/che/docs/stable/administration-guide/extensions-for-microsoft-visual-studio-code-open-source/#selecting-an-open-vsx-registry-instance

Author: Added:

<img width="834" height="312" alt="screenshot-gitlab_cee_redhat_com-2025_08_18-15_15_42" src="https://github.com/user-attachments/assets/b0e387b7-5a93-424c-b77e-f1da8d43ee3f" />

+
[bash,subs="verbatim",options="nowrap"]
----
export CHECLUSTER_NAME="$(oc get checluster --all-namespaces -o json | jq -r '.items[0].metadata.name')" &&
export CHECLUSTER_NAMESPACE="$(oc get checluster --all-namespaces -o json | jq -r '.items[0].metadata.namespace')" &&
export OPENVSX_ROUTE_URL="$(oc get route internal -n openvsx -o jsonpath='{.spec.host}')" &&
export PATCH='{"spec":{"components":{"pluginRegistry":{"openVSXURL":"https://'"$OPENVSX_ROUTE_URL"'"}}}}' &&
oc patch checluster "${CHECLUSTER_NAME}" --type=merge --patch "${PATCH}" -n "${CHECLUSTER_NAMESPACE}"
----

> . Publish Visual Studio Code extensions with the `ovsx` command.

Reviewer: Maybe we should clarify that by default the registry is empty and has no extensions:

<img width="1620" height="580" alt="Screenshot 2025-08-18 at 13 05 46" src="https://github.com/user-attachments/assets/7ae19ca4-6577-4dd4-91b4-6e7457a75d0b" />

Author: Added:

<img width="801" height="146" alt="screenshot-mail_google_com-2025_08_18-15_02_13" src="https://github.com/user-attachments/assets/132838cf-50f5-4512-9d56-468f7d391e83" />

+
With everything configured, the next step is to publish a Visual Studio Code extension from inside the ovsx-cli container.
To do this, you need two pieces of information: the extension `namespace` name (used for publishing), the download URL of the .vsix extension package.
Once you have this information, run the following commands to publish the extension: 
+
.. Retrieve the name of the pod running the Open VSX server:
+ 
[bash,subs="verbatim",options="nowrap"]
----
export OVSX_POD_NAME=$(oc get pods -n openvsx -o jsonpath="{.items[*].metadata.name}" | tr ' ' '\n' | grep ^ovsx-cli)

> ----

Reviewer: @svor looks like actual instructions for publication are missing 🤷

Author: @ibuziuk they are located in the main page `running-the-open-vsx-on-premises.adoc`. If you open https://68a328670f7a316b3a3187ca--eclipse-che-docs-pr.netlify.app/docs/next/administration-guide/running-the-open-vsx-on-premises/#_using_openshift_cli_oc_tool and find step number 9, you should see instructions:

 
<img width="838" height="802" alt="screenshot-gitlab_cee_redhat_com-2025_08_18-15_04_26" src="https://github.com/user-attachments/assets/282b3521-c611-4cdb-8f54-5cb84d7601cc" />

These instructions are identical for upstream and downstream that's why they are not in `snip_che-deploy-open-vsx-using-oc.adoc` snippet (which is upstream specific only)

* Internet connection.
* Installed oc and git tools.
* Installed podman.
* Log in to the OpenShift cluster where a {prod-short} is deployed as an administrator.
+
[TIP]
====

`$ oc login pass:c,a,q[{prod-url}] --username=__<my_user>__`

====

.Procedure

. Create a new OpenShift project for Open VSX.
+
[subs="+attributes,+quotes"]
----
oc new-project openvsx
----
. Clone Open VSX sorces.
+
Get Open VSX sources and navigate to the OpenShift deploy directory:
+
[bash,subs="verbatim",options="nowrap"]
----

> git clone https://github.com/svor/openvsx.git &&

Author: This link will be updated once the instructions have been reviewed and validated.
